File: ruby/getSnaps.rb
# port of getSnapsDir.pl - Jim Bumgardner
#

require 'json'
require 'pp'
require 'flickraw-cached'
require 'optparse'

def makeFlickrPath(photo, suffix)
  return sprintf "http://farm%s.static.flickr.com/%d/%s_%s%s.jpg", 
              photo['farm'],photo['server'],photo['id'],photo['secret'],suffix
end

def makeDirName(id)
  id = id.to_i
  return sprintf 'flickrcache/%03d/%03d/', (id/1000000)%1000, (id/1000)%1000;
end

def makeLocalPath(photo, suffix)
  return makeDirName(photo['id']) + photo['id'] + suffix + ".jpg";
end

def buildDirs(lname)
  dirs = lname.split(/\//)
  dirs.pop
  ldir = '';
  dirs.each do |d|
    break if d =~ /\.jpg/i
    ldir += '/' if ldir != ''
    ldir += d
    `mkdir #{ldir}` if !File.exists?(ldir)
  end
end

$suffix = '_t'
$ifname = 'untitled'
$big = true
$verbose = false
$reverse = false

OptionParser.new do |o|
  o.banner = 'getSnaps.rb [<options>] listfile[.json]'
  o.separator ''
  o.on('-b', '--big', 'Get larger photos') { |b| $big = b; $suffix = '' if $big }
  o.on('-r', '--reverse', 'Reverse order of photos') { |b| $reverse = b; }
  o.on('-v', '--verbose','Verbose messages') { |b| $verbose = b }
  o.on('-h', '--help', 'Display this screen') { puts o; exit }
  o.parse!
end


if (ARGV.length > 0)
  $ifname = ARGV[0]
end
$ifname += ".json" if not $ifname =~ /\.json$/

if !File.exists?($ifname)
  puts "File #{$ifname} does not exist"
  exit
end

puts "Photolist #{$ifname}" if $verbose
photos = JSON.parse(File.read($ifname))
photos.reverse! if $reverse
n = 0
ne = 0
photos.each do |p|
  url_b = makeFlickrPath(p, $suffix)
  l_path = makeLocalPath(p, $suffix)
  if File.exists?(l_path) and File.size(l_path) > 100
    ne += 1
    next
  end
  puts(url_b + " --> " + l_path) if $verbose
  buildDirs(l_path)
  # use curl to capture these...
  `curl -s #{url_b} >#{l_path}`
  n += 1
end
puts "Downloaded #{n} thumbnails" if n > 0 and $verbose
puts "#{ne} thumbs already exist" if ne > 0 and $verbose

File: python/getSnaps.py
# Flickr Image Downloader - supports multiple threads
# make buildDirs lock, so we don't hit it twice...

import json
import requests
import argparse
import sys, os, threading
from subprocess import call
from time import sleep

class DownloaderThread(threading.Thread):
  def __init__(self, idx, photoList):
    threading.Thread.__init__(self)
    self.photoList = photoList
    self.done = False
    self.idx = idx
    self.ne = 0
    self.n = 0

  def makeFlickrPath(self, photo, suffix):
     return "http://farm%s.static.flickr.com/%s/%s_%s%s.jpg" % (
                 photo['farm'],photo['server'],photo['id'],photo['secret'],suffix)

  def makeDirName(self, id):
    id = int(id)
    return 'flickrcache/%03d/%03d/' % ((id/1000000)%1000, (id/1000)%1000)

  def makeLocalPath(self, photo, suffix):
    return self.makeDirName(photo['id']) + photo['id'] + suffix + ".jpg"

  def buildDirs(self, lname):
    dirs = lname.split('/')
    dirs.pop()
    ldir = ''
    for d in dirs:
      if '.jpg' in d:
        break
      if ldir != '':
        ldir += '/'
      ldir += d
      if not os.path.exists(ldir):
        os.mkdir(ldir)

  def downloadImage(self,url, path):
    if os.path.exists(path):
      # print "got",filename
      return
    self.buildDirs(path)
    r = requests.get(url)
    if r.status_code != 200:
      print r.status_code,"on",url
    else:
      imgFile = open(path, "w")
      imgFile.write(r.content)
      imgFile.close()
      print url,"-->",path

  def run(self):
    for p in self.photoList:
      url_b = self.makeFlickrPath(p, suffix)
      l_path = self.makeLocalPath(p, suffix)
      if os.path.exists(l_path) and os.path.getsize(l_path) > 100:
        self.ne += 1
        continue
      self.downloadImage(url_b, l_path)
      self.n += 1
    self.done = True
    # print "Thread",self.idx,"Done"


parser = argparse.ArgumentParser(description='Retrieve thumbnails from Flickr')
parser.add_argument('-b', '--big', default=False, action='store_true', help='Get larger photos')
parser.add_argument('-r', '--reverse', default=False, action='store_true', help='Reverse order of photos')
parser.add_argument('-v', '--verbose', default=False, action='store_true', help='Verbose messages')
parser.add_argument('ifile',  nargs='+', help='file(s) to retreive images from')
args = parser.parse_args()

suffix = '' if args.big else '_t'
verbose = args.verbose
reverse = args.reverse

nbrDownloaded = 0
nbrSkipped = 0

for ifprefix in args.ifile:
  ifname = ifprefix

  if not '.json' in ifname:
    ifname += '.json'

  if not os.path.exists(ifname):
    print "File %s does not exist"
    sys.exit()

  if verbose:
    print "Photo list: " + ifname

  photos = json.loads(open(ifname).read())


  nbrThreads = 8
  partition = len(photos)/nbrThreads + 1
  threads = []
  for i in range(nbrThreads):
    t = DownloaderThread(i,photos[partition*i:partition*i+partition+1])
    t.daemon = True # insures thread dies with main process, if necessary
    t.start()
    threads.append(t)

  allDone = False
  while not allDone:
    allDone = True
    sleep(2)
    for thread in threads:
      if not thread.done:
        allDone = False
        break

  for thread in threads:
    nbrDownloaded += thread.n
    nbrSkipped += thread.ne

if nbrDownloaded > 0 and verbose:
  print "Downloaded %d thumbnails" % (nbrDownloaded)
if nbrSkipped > 0 and verbose:
  print "%d thumbs already exist" % (nbrSkipped)




